html post-processing

sanette opened the pull request:
This PR adds a new feature: 
Post-processing of the html of the OCaml manual to enhance the user experience when viewing the manual and the API in a web browser. The main features are a new style/design (which tries to borrow from the `ocaml.org` website, in view of future integration) and a quick search widget for the API.

After discussions with @gasche and @Octachron and trials in https://github.com/sanette/ocaml-api and https://github.com/sanette/ocaml-tutorial, we thought it would be a good idea to integrate these in the general build of the OCaml manual.

This adds a new rule:
`make web`

Please see the short description in `html_processing/README.md`.

The output of this is a new directory `manual/manual/webman` containing a standalone browsable 'web site' in two subdirectories: `api` and `manual`.

Discussions about this have been happening in https://discuss.ocaml.org/t/suggestions-for-ocaml-documentation/4504 and 
https://discuss.ocaml.org/t/new-pages-for-ocaml-api/4720, which was very useful in the development of the project.

sanette committed:
add html post-processing

Thanks to @gasche and @Octachron for help in structuring this.

nojb commented:
A couple of high-level questions without having looked at the code:
- Is the result of the postprocessing accessible anywhere? It would help reviewing the PR.
- Why *post*-processing? Is it the case that it would be too complicated to generate the "right" HTML from the start?

sanette commented:
> A couple of high-level questions without having looked at the code:
> 
> * Is the result of the postprocessing accessible anywhere? It would help reviewing the PR.

yes, it's essentially identical to https://sanette.github.io/ocaml-tutorial/4.10/index.html

> * Why _post_-processing? Is it the case that it would be too complicated to generate the "right" HTML from the start?

Yes, exactly. This is a good question, which was discussed for instance in https://discuss.ocaml.org/t/suggestions-for-ocaml-documentation/4504
Part of the html could be obtained by rewriting the TeX source and/or the Hevea program, but it's not clear that this was a viable option, in terms of complexity. Other parts, like the quick-search widget, obviously require an add-on.

I agree that ideally we don't want to pile up too many layers, but in the end we were not able to find a reasonable alternative. (And at least in terms of build times, if this matters, this one is orders of magnitude faster than TeX+hevea.)

sanette commented:
> yes, it's essentially identical to https://sanette.github.io/ocaml-tutorial/4.10/index.html

If you think it would help having an *exact* mirror, I can set up one too.
The differences should be
* OCaml version 4.11 instead of 4.10
* An explicit link to the "compiler API" in the Manual Table of Contents
* The OCaml logo in the API section links back to the Manual table of contents, instead of ocaml.org
* somewhat simplified CSS
* the scripts themselves are simplified because they don't try to process older OCaml versions < 4.10

dbuenzli commented:
> Part of the html could be obtained by rewriting the TeX source and/or the Hevea program, but it's not clear that this was a viable option, in terms of complexity.

I think in the end you will end up with something much less complex but also easier to contribute to in the long term. It may mean a deep dive and cleanup to manual sources and LaTeX definitions but I think this would be the sane way to go. As I wrote [here](https://discuss.ocaml.org/t/suggestions-for-ocaml-documentation/4504/37) HeVeA can perfectly be made to generate modern and well `class`ifed HTML. This may however entail spending a bit of quality time with `HeVeA`'s [nice manual](http://hevea.inria.fr/doc/index.html) though.

To encourage contributions to the manual it should be made easier to tweak and build. Piling layers and layers of processing doesn't help in my opinion. Also I find the use of `sass` here entirely unnecessary here now that you have `calc` and variables in CSS; it just make it harder for contributors (have to understand sass, have to install the tools to process them, etc.). 

Having a base line manual with clean and properly classified markup would also make it easier to blend the manual in personalized documentation systems like `odig` (see for the example the link to the ocaml manual link on [this page](https://b0-system.github.io/odig/doc@odig.light/)). At the moment addressing some of the elements of the manual for styling can be [horrendous](https://github.com/b0-system/odig/blob/master/themes/manual.css#L191) and even with that the results are not always as satisfying as they could be.

gasche commented:
@Octachron did integrate some of @sanette's experiments into the existing manual build system:
- #9089, source code highlighting for manual examples
  (it's using lstlisting and the result is not as nice as @sanette's HTML-only solution)
- #9101, visible anchors when hovering on a heading

The proposal to merge some other parts of @sanette's work as post-processing comes from a more pragmatic cost-benefit analysis:
 - Ideally having no postprocessing would be nicer, but integrating each postprocessing pass in the existing pipeline can be a lot of work, and we thought that this work was unlikely to happen in practice for many of the remaining transformations (it was only @Octachron doing the work and we have better uses of his time). Of course, we are always happy to move more things out of the postprocessing phase if there are contributions to do it.
- @sanette's work started as an experiment, but our impression is that it is surprisingly clean in terms of code, and something we could see ourselves maintaining in the future. For me personally this is just an impression, I have not reviewed the code, but hopefully determining this (maintainability of the post-processing logic) will be an outcome of the PR review process.
- Some of the "manual evolution" plans being considered (switching to odoc: strong support, medium-term change) (switching to AsciiDoc or some other markup language: less clear support, longer-term change) could obsolete some parts of the current manual-production pipeline (ocamldoc for library documentation, hevea for manual production), so it is not clear that we want to invest more time in improving those tools specifically.

So: yes postprocessing adds complexity, but it may still be a pragmatic choice (this was our impression with @Octachron, and he (and myself to some extent) have been the most active contributors on the manual-pipeline front lately). If there are changes we can suggest to make the post-processing stack simpler, it's a good idea to do them as part of the review process, and I'm definitely curious about maintainability assesments. But I don't think we should reject post-processing in principle. (Note: my opinion evolved on this, initially I was rather against HTML post-processing for much of the same reasons.)

sanette commented:
Concerning syntax highlighting, in earlier versions I was using @zoggy 's`higlo`, see for instance https://sanette.github.io/ocaml-tutorial/4.09/coreexamples.html . But as soon as @Octachron implemented it upstream in the LaTeX source, it was clear to me that I had to remove it from the post-processing. (Maybe Higlo's results were somewhat nicer, but on the other hand some constructs were not recognized and it was not maintained anymore so anyway it was not a very good solution.) The current proposal does not use `higlo` anymore.

This is a good example of the pragmatic approach. Once features are integrated into a deeper layer, we can easily remove them from post-processing.

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+open Soup
+open Printf
+
+let debug = Array.to_list Sys.argv
+            |> List.mem "quiet"
+            |> not
```
```suggestion
let debug = not (Array.mem "quiet" Sys.argv)
```

dra27 replied:
I'm a middling fan of the `|>` operator when the code really resembles a pipeline, but this I think qualifies as overuse (it's like using the ancient calculators at my school...!)

sanette replied:
Sure. Let's do this.

dra27 committed:
Missing .gitignore entries

dra27 committed:
Allow compilation without opam

Adds a duniverse target which clones and compiles Dune and clones the
dependencies. The process assumes that the compiler in the tree has been
installed and is in PATH.

dra27 committed:
Test manual compilation in Travis

dra27 committed:
Fix web target

dra27 committed:
Clean-up Dune build in the tree

dra27 committed:
Fix incorrect clean target in manual Makefile

dra27 commented on manual/manual/html_processing/README.md:
```diff
@@ -0,0 +1,57 @@
+search widget for the API.
+
+The process will create the `../webman` dir, and output the new html
+files (and assets) in `webman/manual` (the manual) and `webman/api` (the
```
```suggestion
files (and assets) in `../webman/manual` (the manual) and `../webman/api` (the
```

dra27 commented on manual/manual/html_processing/README.md:
```diff
@@ -0,0 +1,57 @@
+files (and assets) in `webman/manual` (the manual) and `webman/api` (the
+API).
+
+For processing the manual/api of OCaml versions <= 4.10, see
+https://github.com/sanette/ocaml-tutorial and
+https://github.com/sanette/ocaml-api
```
This doesn't need to go here

sanette replied:
OK, I'll remove this.

dra27 commented on manual/manual/html_processing/README.md:
````diff
@@ -0,0 +1,57 @@
+make
+```
+
+The scripts should run with `ocaml >= 4.05`.  You also need a working
````
```suggestion
You need a working
```

dra27 commented on manual/manual/html_processing/README.md:
```diff
@@ -0,0 +1,57 @@
+
+## Debug
+
+Look at the `Makefile`: the `quiet` keyword is optional. Remove it to
```
This should be done with a `DEBUG` variable in `Makefile` so that you can issue `make DEBUG=1` (in general, avoid editing files in order to investigate things as the changes may be accidentally committed)

sanette replied:
I can try to do this but my knowledge of Makefile syntax is close to zero...

sanette replied:
done

dra27 commented on manual/manual/html_processing/js/scroll.js:
```diff
@@ -0,0 +1,66 @@
+
+// Get absolute y position of element.
+// modified from:
+// https://www.kirupa.com/html5/get_element_position_using_javascript.htm
```
Could you note effective [CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) licence on the original and reference https://forum.kirupa.com/t/get-an-elements-position-using-javascript/352186/3

sanette replied:
Done. But, on second thought, sincne the code is actually quite different and strip down to almost nothing, do we really have to reference the "source"?

dra27 replied:
If you're confident to attribute it as yours, that's fine!

dra27 commented on manual/manual/html_processing/js/scroll.js:
```diff
@@ -0,0 +1,66 @@
+// Smooth scrolling only for near targets
```
If I interpret correctly, this JavaScript is entirely standalone and, more importantly, optional?

dra27 replied:
If accepted, it needs a full copyright header

sanette replied:
exactly, it's entirely standalone and optional

dra27 commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+// Searching the OCAML API.
```
Full copyright header needed

dra27 replied:
What's the impact of _not_ including this JavaScript file?

sanette replied:
The search functionality is based on this javascript file. If it is not included, search won't work, but apart from this (and a few js errors in the console) the site will be normally browsable

dra27 replied:
OK - my only intention with it is to be clear for files in another language what the maintenance impact of them is - it sounds like we'd be cleanly able to remove them, if necessary (I don't think it'd be likely, though!)

sanette replied:
I have added a remark about this in the README

dra27 commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+    return acc + a;
+}
+
+// NOT used
```
So does it need to stay?

sanette replied:
I tend to leave things that I think may be useful if I change my mind on a particular feature... but in the end they are rarely useful, so let's remove this one.

dra27 commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+//console.log (hasSubString ("ell", line));
+
+//console.log(GENERAL_INDEX.filter(function (line) {
+//    return (hasSubString("iter", line)); })[0]);
```
```suggestion
```

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+            |> List.mem "quiet"
+            |> not
+
+let pr = if debug then print_endline else fun _ -> ()
```
I would call this function `dbg` or `debug`, not `pr`. I think you'll find this more useful:

```ocaml
let dbg =
  let printf = Printf.(if debug then kfprintf else ikfprintf) in
  let flush =
    if debug then
      fun ch -> output_char ch '\n'; flush ch
    else
      ignore
  in
  fun fmt -> printf flush stdout fmt
```

as you can now have printf for free in every call (you use `sprintf` and `^` a lot below). The only thing you can't do is `dbg file` which now becomes `dbg "%s" file`

sanette replied:
I didn't have in mind to write a powerful debug function for my basic debug messages, but I'm ok to use your suggestion. How do I acknowledge you for this?

dra27 replied:
No acknowledgement needed. I'd heartily recommend using one like this in general, though - you win the second you have more than one variable to print, and it can also occasionally be a debugging nightmare when the messages aren't being flushed immediately!

sanette replied:
What would be the difference with this:
```ocaml
let dbg s = Printf.ksprintf (fun s -> if debug then print_endline s) s;
```
(I guess: performance?)

dra27 replied:
Not sure about performance per se, but it's superior when you start using `%a` - the key thing with `ikfprintf` (and `ifprintf`) is that it doesn't actually format the string. You also get to stay in the `fprintf` class of `%a` functions (so `out_channel -> 'a -> unit` rather than `unit -> 'a -> string`).

This does nothing and returns immediately:
```ocaml
Printf.ifprintf stdout "%a" (fun _ _ -> Unix.sleep 5) 42
```
this "does the same nothing", but takes 5 seconds:
```ocaml
Printf.ksprintf ignore "%a" (fun _ _ -> Unix.sleep 5; "") 42
```
That's useful if, for example, what you want to print in the debug statement is computationally intensive, but remember that even `string_of_int` (i.e. `%d`) is not free - `Printf.ikfprintf stdout "%d" 42` doesn't call `string_of_int` (and so also doesn't allocate any string to hold the result)

sanette replied:
works great, thanks

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+
+let with_dir = Filename.concat
+
+let process_dir = "."
```
```suggestion
let process_dir = Filename.current_dir_name
```
(pedantically)

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+let process_dir = "."
+
+(* Output directory *)
+let web_dir = "../webman"
```
I haven't actually tested this on Windows, but `Filename.concat` is nicer than hard-coded forward-slashes.
```suggestion
let web_dir = Filename.concat parent_dir_name // "webman"
```
(and for all hard-coded `../` below)

sanette replied:
ok (you meant `let web_dir = Filename.parent_dir_name // "webman"`)

dra27 replied:
Oops, yes of course!

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+
+let pr = if debug then print_endline else fun _ -> ()
+
+let with_dir = Filename.concat
```
```suggestion
let ( // ) = Filename.concat
```

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+
+(**** utilities ****)
+
+let do_option f = function
+  | None -> ()
+  | Some x -> f x
+
+let map_option f = function
+  | None -> None
+  | Some x -> Some (f x)
+
+let flat_option f = function
+  | None -> None
+  | Some x -> f x
```
These are `Option.iter`, `Option.map` and `Option.bind` respectively.

sanette replied:
yes, but I usually develop with my ocaml-system switch, which is currently 4.05, so if possible I would prefer keep them (or rename)

dra27 replied:
If it's being upstreamed, it's for `trunk` OCaml... it sounds like the `duniverse` target I added would be useful to you directly! If you `./configure --prefix $(pwd)/install; make -j world.opt; make install; PATH=$(pwd)/install:$PATH make -C manual/manual/html_processing duniverse` and you now have the trunk compiler to develop with (just do the `PATH=` bit before `make` in `html_processing` or export it)

sanette replied:
isn't it useful to be able to build the trunk manual with an older ocaml version?
(I mean, without having to install everything with the duniverse target)
I thought so, but if you think that's not useful, I will use the Option module as you suggested.

dra27 replied:
Not for core development - it's mild useful to be able to build it with _recent_ OCaml, but make -j world.opt doesn't take that long!

sanette replied:
OK, I'm now using Option

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+
+let string_of_opt = function
+  | None -> ""
+  | Some s -> s
```
```suggestion
let string_of_opt = Option.value ~default:""
```

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  if Sys.command (sprintf "cp %s %s" file dst) <> 0
+  then failwith ("Could not copy " ^ file)
+
+let sys_mkdir dir =
+  if Sys.command (sprintf "mkdir -p %s" dir) <> 0
+  then failwith ("Could not create directory" ^ dir)
```
Use `Sys.remove`

sanette replied:
why `Sys.remove` ?

dra27 replied:
Oops, because apparently I can't read! There will shortly be `Sys.mkdir`

sanette replied:
In view of your remarks, finally I am removing all sys_* commands and instead execute them in the Makefile.
Btw this is probably done in a very naive way in the Makefile, feel free to improve.

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  if Sys.command (sprintf "mkdir -p %s" dir) <> 0
+  then failwith ("Could not create directory" ^ dir)
+
+let sys_mv file dst =
+  if Sys.command (sprintf "mv %s %s" file dst) <> 0
+  then failwith ("Could not move " ^ file)
```
Use `Sys.rename`

sanette replied:
I was using `Sys.rename` first, but unfortunately it didn't work when moving from one partition to another.

dra27 replied:
Oh, I see it's an actual move - it would still be much more portable to do a copy followed by `Sys.remove` in that case

sanette replied:
cf https://github.com/ocaml/ocaml/pull/9755#discussion_r479674453

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  prepend_child nav vnum
+
+(* Some wrappers around linux system commands*)
+let sys_cp file dst =
+  if Sys.command (sprintf "cp %s %s" file dst) <> 0
+  then failwith ("Could not copy " ^ file)
```
I'd lift `Ocamltest_stdlib.Sys.copy_file` from https://github.com/ocaml/ocaml/blob/trunk/ocamltest/ocamltest_stdlib.ml#L143-L159 rather than this

sanette replied:
https://github.com/ocaml/ocaml/pull/9755#discussion_r479674453

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  let src = with_dir process_dir src in
+  if Sys.command (sprintf "sass %s > %s" src dst) <> 0
+  then sprintf "Could not compile %s to %s. Is sass installed?" src dst
+       |> failwith
```
Surely this can be done as a `dune` rule or in `make`?

sanette replied:
yes, probably. Would it be better?

dra27 replied:
Well, it allows the build system to minimise work - you'll only call sass if the files have actually changed

sanette replied:
done

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  then sprintf "Could not compile %s to %s. Is sass installed?" src dst
+       |> failwith
+
+(* Detect OCaml version from version.tex *)
+let find_version () =
+  let versiontex = "../version.tex" in
+  Scanf.bscanf (Scanf.Scanning.from_file versiontex)
+    "\\def\\ocamlversion{%s@}" (fun a -> a)
```
You can cut out the middle man, as you're in the tree - it's the first line of `../../../VERSION`!

sanette replied:
thanks ;)

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+  Scanf.bscanf (Scanf.Scanning.from_file versiontex)
+    "\\def\\ocamlversion{%s@}" (fun a -> a)
+
+(*
+   Local Variables:
+   compile-command:"dune build"
+   End:
+*)
```
What's this?

sanette replied:
it's setting my emacs local variables;) I can then quickly run "dune build" with C-c C-c

dra27 replied:
Ah, fair enough!

dra27 commented on manual/manual/html_processing/src/common.ml:
```diff
@@ -0,0 +1,123 @@
+open Soup
```
Copyright header needed

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+(* Post-processing the HTML of the OCaml API.
```
Copyright header needed

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+let compiler_libref = ref false
+(* set this to true to process compilerlibref instead of libref *)
+
+let libref = ref ""
```
This is not used outside `set_dirs`, so should not be global

sanette replied:
good catch

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+open Printf
+open Common
+
+let compiler_libref = ref false
```
I think this code would be clearer if instead of these refs and peppering of `if !compiler_libref` in various places you defined a record `config` or `build` containing the required information (`{src_src = ..; dst_dir = ..; title = "Compiler "; prefix = "../"`)

sanette replied:
done

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+    ) []
+
+
+(* Fast version of make_index using Scanf *)
```
You don't seem to use either version? What difference in speed are we talking here?

sanette replied:
Scanf version is about 20x faster

sanette replied:
following your suggestion, I finally opted for the slow version, but with a cache to speed up a little bit (on my laptop, the slow version takes 18 sec with caching, 50 sec without, and the fast version takes 1 sec).

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+(* HTML code for the search widget. We don't add the "onchange" event because it
+   forces to click twice to an external link after entering text. *)
+let search_widget with_description =
+  sprintf "<div class=\"api_search\"><input type=\"text\" name=\"apisearch\" id=\"api_search\"
+	 oninput    = \"mySearch(%b);\"
+         onkeypress = \"this.oninput();\"
+         onclick    = \"this.oninput();\"
+	 onpaste    = \"this.oninput();\">
+<img src=\"search_icon.svg\" alt=\"Search\" class=\"svg\" onclick=\"mySearch(%b)\">%s</div>
+<div id=\"search_results\"></div>" with_description with_description
```
If you use `{|<div class="api_search"...|}` then you want to quote all the double-quotes.

sanette replied:
ok, done

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+	 onpaste    = \"this.oninput();\">
+<img src=\"search_icon.svg\" alt=\"Search\" class=\"svg\" onclick=\"mySearch(%b)\">%s</div>
+<div id=\"search_results\"></div>" with_description with_description
+    (if with_description
+     then {|<span class="search_comment">(search values, type signatures, and descriptions - case sensitive)<div class="search_help"><ul><li>You may search bare values, like <code>map</code>, or indicate the module, like <code>List.map</code>, or type signatures, like <code>int -> float</code>.</li><li>To combine several keywords, just separate them by a space - except if you want to search type  ...
+     else "")
```
I'd bind this:

```ocaml
let search_description = if with_description ... else "" in
sprintf {| ... |} with_description with_description search_description
```

sanette replied:
ok, done

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+
+let process ?(search=true) ~version file out =
+
+  sprintf "Processing %s..." file |> pr;
```
```suggestion
  dbg "Processing %s..." file;
```
(and _all_ other instances!)

sanette replied:
done

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+  if search then begin
+    let head = soup $ "head" in
+    create_element "script" ~attributes:["src","search.js"]
+    |> append_child head;
```
For me, this is a good example of `|>`: "create an element **and** append it as a child of the head"

sanette replied:
I agree but sometimes I also like to do eg:
```
mylist |> iter (function a -> .....)
```
especially when the code of the function is long and would hide the list if we put it at the end

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+
+  (* Add favicon *)
+  let head = soup $ "head" in
+  {|<link rel="shortcut icon" type="image/x-icon" href="|} ^ favicon () ^ {|">|}
+  |> parse
+  |> append_child head;
```
```suggestion
  parse ({|<link rel="shortcut icon" type="image/x-icon" href="|} ^ favicon () ^ {|">|})
  |> append_child head;
```

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+  |> do_option delete;
+
+  (* Create TOC with H2 and H3 elements *)
+  (* It would be much simpler to insert only H2 *)
```
What's the relevance of this comment?

sanette replied:
relevant only for me I guess ;)  I changed it

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+
+  descendants body
+  |> elements
+  |> iter (fun h -> match name h with
```
Can't you use `fold` here instead, and have h3_open as the value you're folding?

sanette replied:
OK, then I can even fold over `li_current,h3_current` and remove `h3_open`

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+  let outch = open_out file in
+  output_string outch "var GENERAL_INDEX = [\n";
+  index
+  |> List.iter (fun (mdule, value, infolist, signature) ->
```
Matching pairs here:
```suggestion
  |> List.iter (fun ((mdule_l, mdule_r), (value_l, value_r), (infolist_l, infolist_r), signature) ->
```
...

sanette replied:
ok

dra27 commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,414 @@
+  index
+  |> List.iter (fun (mdule, value, infolist, signature) ->
+      fprintf outch {|["%s", "%s", "%s", "%s", "%s", "%s", "%s", "%s"],|}
+        (fst mdule) (snd mdule) (fst value) (snd value) (fst infolist) (snd infolist)
```
... will eliminate all the `fst` and `snd` calls here

sanette replied:
done

dra27 commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,370 @@
+  part
+
+
+(* will be automatically updated *)
```
how/when?

sanette replied:
you're right, this was a copy-paste from the `process_api.ml`, but in the current case, it was more complicated to try to update this string, and I finally gave up. Let's remove this comment.

dra27 replied:
It's just `\number\year` in the TeX files - why not just use `Unix.gettimeofday`?

dra27 commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,370 @@
+let load_html file =
+  pr file;
+  (* First we perform some direct find/replace in the html string. *)
+  (* Warning charset = ascii for 4.05 and utf8 for >=4.09.  But the original
+     html is kind of buggy because 4.05 uses &#XA0; for non-breaking space in
+     us-ascii encoding, although &#XA0; is latin encoding...  Finally we decide
+     to force utf8 encoding.  *)
```
Not needed for here, therefore

sanette replied:
right, I'll remove this comment

dra27 reviewed (commented):
I've taken the liberty of pushing some build system changes/fixes directly to your branch which you are welcome either to take or, of course, reject. I'm not sure if we should continuously test the manual in Travis, necessarily.

This is a fairly high-level review:
- Neither the slow nor fast index code is used - unless the lambdasoup is 1000 times slower, I'd opt for slow-and-correct (the manual itself takes a while to compile, and it's not exactly performance critical code!). Maintaining two versions of a generator is definitely not viable.
- I think this may already be the case, but it would be good to have a clear way of eliminating the JavaScript, just to preserve the option of losing it in the future.
- Given that you already pull in a couple of dependencies, I think it would be better to pull in ocaml-re and not use Str, FWIW

In many respects, keeping this as post-processing is initially a good thing, as it vastly lowers the maintenance burden as the manual isn't exactly unusable without it.

sanette commented:
@dra27 thanks for the review!
I don't have much time to work on this right now, but I just want to mention that the fast index is used by default by the function `process_index`

dra27 commented:
Ah, thanks - I was being blind yesterday, and could only see the one guarded by the unused `makeindex` parameter. In which case, I see next to no difference in run-time between the two versions on my system (the whole thing takes 8-10s which is still considerably faster than a full build of the manual). So I'd definitely go with the slow-and-correct soup version!

sanette committed:
Update manual/manual/html_processing/src/common.ml

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette committed:
Update manual/manual/html_processing/README.md

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette commented:
on my laptop, running `time dune exec src/process_api.exe makeindex quiet`,
I get about 18 sec for the slow version and less than 1 sec for the fast version.
I agree this is negligible compared to the whole build of the manual! But when I develop in the `html_processing` dir, of course I don't have to build the whole thing for each modification, and the factor 18 makes a difference...
Another remark is that the slow version is a bit deprecated: some recent features have been implemented in the fast version only (searching by type signature). Of course this can be corrected, if you really think it is better.

sanette committed:
Update manual/manual/html_processing/README.md

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r463601452

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r463553488

et aussi

https://github.com/ocaml/ocaml/pull/9755#discussion_r463554035

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r463554035

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r463554136

sanette committed:
Update manual/manual/html_processing/js/search.js

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette committed:
Update manual/manual/html_processing/src/common.ml

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette committed:
Update manual/manual/html_processing/src/common.ml

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r463559857

dra27 commented:
The main thing is just to have one version - instinctively, one of prefers the principled approach with lambdasoup, although in practice it's probably not much less brittle than the Scanf version! For speed of development, couldn't you work with a fake smaller stdlib?

Anyway, from a maintenance perspective - having picked one - what would also be helpful would be high-level comments on the functions for the transformations being performed and why. What I mean by that is something along the lines of:

```ocaml
(* The file we're processing has this structure:
<tag1>
  <tag2 attr="something-we-track">
     <tag3 id="function-name">Stuff we care about</tag3>
      ...
  </tag2>
</tag1>
which we want to transform to be:
<amazing-tag>
  <amazing-tag2>something-we-track</amazing-tag2>
  <amazing-tag3 id="function-name">Transformed stuff we care about</amazing-tag3>
</amazing-tag>
*)
```

sanette commented:
> For speed of development, couldn't you work with a fake smaller stdlib?

I don't quite understand this, could you elaborate?

dra27 commented:
I mean if indexing lots of files is slow, can’t you temporarily make the thing you’re indexing smaller?

sanette commented:
> * I think it would be better to pull in ocaml-re and not use Str,

Is ocaml-re supposed to replace Str in the future?
I couldn't find an online doc for ocaml-re

sanette committed:
move DEBUG to Makefile

sanette committed:
implement some of @dra27 suggestions + correct js

sanette committed:
move all shell stuff (cp, rm...) into the Makefile

sanette committed:
add copyrights

sanette committed:
complete lambdasoup version of make_index

However, lambdasoup version takes 56sec while the Scanf version takes 1sec...

sanette committed:
improve make_index

sanette committed:
remove global variables

Octachron commented:
We are thinking of moving `Str` out of the standard distribution at some point in the future (ocamldoc is still an user of `Str`, so that this will not happen before the switch to odoc).

Octachron commented on manual/manual/html_processing/scss/style.scss:
```diff
@@ -0,0 +1,1089 @@
+	    /* this should be the column with the type */
+	    width: 60%;
+	    /* not optimal, see: file:///home/san/prog/ocaml/ocaml-api/dune-version/docs/Format.html#symbolic
+	    but leaving it automatic is not always good either: see: file:///home/san/prog/ocaml/ocaml-api/dune-version/docs/Arg.html */
```
Those paths should be relative.

sanette replied:
ok, done

sanette committed:
use ocaml-re instead of str

sanette commented:
> We are thinking of moving `Str` out of the standard distribution at some point in the future (ocamldoc is still an user of `Str`, so that this will not happen before the switch to odoc).

ok, I changed to Re.Str and removed str dependency. I suppose that the duniverse target added by @dra27 must be updated, then.

sanette committed:
make paths relative in comments

Octachron commented:
The rendering of variants is off with firefox. They look like:

```
type complex64_elt = 
  |
Complex64_elt
```

Octachron commented on manual/manual/html_processing/scss/manual.scss:
```diff
@@ -0,0 +1,292 @@
+    }
+    nav.toc{
+	@include nav-toc-mobile;}
+}
```
In this mode, it is impractical to go to the next page when reading linearly: the table of the content is at the top of the page and there is no navigation element at the bottom of the page.

sanette replied:
what do you mean by next page? do you mean next section? I agree this is not very practical. I'm not a big user of the mobile interface, what would users expect? I guess, a button to open the TOC ?

Octachron replied:
I meant the next section or page in the linear order of the manual. Maybe a link to the TOC anchor, previous and next pages? In fact, the next and previous links could be useful in the normal mode. Currently going to the next page takes between two and three actions away: either a click to the OCaml manual page, a search in the Toc, and then a click if we cross a part boundary; or it requires to go back to the top of page, and click on the next section. Having an one click shortcut might be worthwhile?

sanette replied:
apparently, `odoc` doesn't handle this either. I have some ideas but this would require some more javascript.

Octachron replied:
I think this is mostly relevant in the manual (non-api) part of the documentation (where currently I can use the hevea arrows).

sanette replied:
my opinion was that reading the whole manual in a linear way was not useful, but maybe I'm wrong.
To me, the "next/previous" links had no intelligent meaning: when you are at the end of a chapter, you have no clue about what you will discover by clicking on these links, that's why I removed them.

sanette replied:
what one could do is to replace "next/previous" by the title of the chapters they link to

Octachron replied:
The manual is written to be readable in a linear way .
If we could have link with the names of the previous and next chapter (at the bottom to not repeat the TOC) that would be ideal in my opinion.

sanette replied:
great, I'll try to do this then

sanette replied:
done

Octachron replied:
This works very well, thanks!

sanette replied:
thanks! I agree this is a nice improvement

Octachron commented on manual/manual/html_processing/scss/manual.scss:
```diff
@@ -0,0 +1,292 @@
+    }
+    td {
+	background:linear-gradient(to left,#fff 0,#ede8e5 100%);
+	border:1px solid #ccc;
```
Looking at the table in lablexamples.html, I am not really convinced by the gradient here.

sanette replied:
ok, what would you suggest instead?
this had the advantage of clearly separating the columns without resorting to a strong separator

Octachron replied:
Maybe we could start with a simple light flat color?

sanette replied:
ok, I'll do that

sanette replied:
done

sanette commented:
> The rendering of variants is off with firefox. They look like:
> 
> ```
> type complex64_elt = 
>   |
> Complex64_elt
> ```

weird, I see it correctly with my Firefox and also with Chrome
![Screenshot_20200904_174845](https://user-images.githubusercontent.com/7536536/92258739-4b7bb800-eed7-11ea-81a8-1e89a99cb255.png)

Octachron commented:
The newlines seems to only appear at zoom level 90-130%... The issue might only exists on my side.

sanette commented:
> The newlines seems to only appear at zoom level 90-130%... The issue might only exists on my side.

oh I see what you mean, indeed something strange happens at those zoom levels. I have to look into this. Maybe this is a Firefox issue because I have no problem at any zoom level with chrome or qutebrowser

Octachron commented on manual/manual/html_processing/scss/manual.scss:
```diff
@@ -0,0 +1,292 @@
+
+.ocaml{ 
+    background:linear-gradient(to left,#fff 0,#f3f7f6 100%)
+}
```
I have the impression that this this background property is always shadowed by the one in `pre`. Am I missing something?

sanette replied:
I'm not sure either. Maybe I had to add this in order to deal with older ocaml manual versions, I don't remember

Octachron replied:
Then, it is probably better to remove it or at least use the same gradient that in the pre case.

sanette replied:
you're perfectly right. I've just checked that every text inside an "ocaml" class is in fact also inside a "pre" class. So I can remove this statement.

sanette committed:
table background color

https://github.com/ocaml/ocaml/pull/9755#discussion_r483708459

sanette committed:
css for variants

https://github.com/ocaml/ocaml/pull/9755#issuecomment-687234444

sanette commented:
> The rendering of variants is off with firefox. They look like:
> 
> ```
> type complex64_elt = 
>   |
> Complex64_elt
> ```

is this better now ?

Octachron commented:
This looks fixed.

sanette committed:
add navigation links at bottom of page

https://github.com/ocaml/ocaml/pull/9755#discussion_r483705906

dra27 commented on manual/manual/html_processing/Makefile:
```diff
@@ -1,13 +1,84 @@
+#JS_FILES := $(WEBDIRAPI)/scroll.js $(WEBDIRAPI)/search.js $(WEBDIRCOMP)/scroll.js $(WEBDIRCOMP)/search.js $(WEBDIRMAN)/scroll.js
+
+# There must be a more clever way
+$(WEBDIRAPI)/%.js: js/%.js
+	cp $< $@
+
+$(WEBDIRMAN)/%.js: js/%.js
+	cp $< $@
+
+$(WEBDIRCOMP)/%.js: js/%.js
+	cp $< $@
```
```suggestion
define GEN_COPY_RULE
$(WEBDIR$(1))/%.js: js/%.js
	cp $$< $$@
endef

$(foreach target, API MAN COMP, $(eval ($call GEN_COPY_RULE,$(target))))
```
but to be honest I'd leave it as-is!!

sanette replied:
I see, thanks. Then let's leave it as is.

dra27 commented on manual/manual/html_processing/Makefile:
```diff
@@ -1,13 +1,84 @@
+IMG_FILES1 := $(IMG_FILES0) search_icon.svg
+IMG_FILES := $(addprefix $(WEBDIRAPI)/, $(IMG_FILES1)) $(addprefix $(WEBDIRCOMP)/, $(IMG_FILES1)) $(addprefix $(WEBDIRMAN)/, $(IMG_FILES0)) 
+
+# There must be a more clever way
```
Could possibly use the trick above...

sanette replied:
in view of your previous comment, let's leave it as it for now

Octachron commented:
There are few issues in the language reference part (and only this part):
- The chapter preamble is repeated on each page (e.g. https://sanette.github.io/ocaml-tutorial/4.10/locallyabstract.html and https://sanette.github.io/ocaml-tutorial/4.10/privatetypes.html)
- The chapter introduction is not followed by the TOC. Compare https://sanette.github.io/ocaml-tutorial/4.10/extn.html with https://ocaml.org/releases/4.10/htmlman/extn.html . Even if the TOC would be redundant currently, the chapter page feels empty without those links.

sanette commented:
>     * The chapter preamble is repeated on each page (e.g. https://sanette.github.io/ocaml-tutorial/4.10/locallyabstract.html and https://sanette.github.io/ocaml-tutorial/4.10/privatetypes.html)

this was done on purpose in order to better integrate in the page design the sections that consist of separate files, why do you think it is a bad thing?

sanette commented:
>     * The chapter introduction is not followed by the TOC. Compare https://sanette.github.io/ocaml-tutorial/4.10/extn.html with https://ocaml.org/releases/4.10/htmlman/extn.html . Even if the TOC would be redundant currently, the chapter page feels empty without those links.

it looked ok to me, I thought it was clear enough that you had to choose an entry from the side bar. I can repeat the toc in the main section if you prefer, but wouldn't it be confusing to have twice the same toc?

Octachron commented:
First, the preamble for chapter 7 is far too long to be repeated. But generally, I don't see the point in repeating the chapter introduction before each section. I don't want to reread the description of the chapter at each section. Do you have some examples of online books using this kind of repetitions?

For the toc, I don't think that this a problem to repeat the information in a slightly different way twice. Moreover having the toc in the flow of the text conveys the information that the following sections are less cohesive than the previous part of the manual (the tutorial). More importantly, one should not need to go back to the toc to navigate to the first children of a page: this property is currently respected everywhere except in part 2.

Octachron commented on manual/manual/html_processing/scss/manual.scss:
```diff
@@ -0,0 +1,311 @@
+
+.pre,pre{
+    background:linear-gradient(to left,#fff 0,#ede8e5 100%)
+}
```
Note that before merging this PR, we should discuss this gradient. I really dislike it when reading code block. In particular on wide examples, this gradient alters the contrast and the visibility of the code depending on the position in the code block. Should we open a poll?

sanette replied:
a poll could be useful if we restrict participants to beginners, since after all the "tutorial" part is for them.
But maybe it's not worth the trouble. My mission for the OCaml foundation is to make the website look more modern and attractive (and useful of course), and gradients are "back in fashion" since a couple of year; but if you think that's not good for code block, no problem I will remove them

Octachron replied:
My issue is really only with this gradient on a large text-area. The other gradient looks fine.

sanette replied:
It's not easy to tell which ones are large just using the CSS. I did remove gradients for`.pre`. Is this OK with you or did you mean anything else?
(in which case it might be easier for the discussion to explicitly mention which page)

Octachron commented on manual/manual/html_processing/images/colour-logo-gray.svg:
```diff
@@ -0,0 +1,147 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
I don't think that we should commit any images in the tree: for local build, they should be downloaded from the ocaml.org source in the makefile, and once deployed to ocaml.org, the manual should use the resource here. 

There is also an issue with the favicon not appearing in the manual part, and only in the API part of the documentation.

Commenting on this favicon, there are few  more issues:
- there is`#` overlaid on the usual logo
- it is far too big, 130K compared to the 1K favicon served by ocaml.org
It seems really better to get back to `ocaml.org` favicon here.

sanette replied:
there was a discussion about the favicon here: https://github.com/sanette/ocaml-api/issues/2

Octachron replied:
Having a distinct favicon for the experimental or local version is an interesting idea. However, this argument does not seems to apply for the production version. (And if we have a local version of the favicon, I would vote for a svg one since this means text data rather than binary).

sanette replied:
> I don't think that we should commit any images in the tree: for local build, they should be downloaded from the ocaml.org source in the makefile, and once deployed to ocaml.org, the manual should use the resource here.

OK, then I've openned this PR: https://github.com/ocaml/ocaml.org/pull/1166
For the search icon, I will use https://ocaml.org/img/search.svg

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+function isSubString (sub, s) {
+    // s = s.toLowerCase();
+    // sub = sub.toLowerCase();
+    // case sensitive is better for OCaml modules.
```
The dead code before the comment can be removed.

sanette replied:
done

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+// for each sub we compute the minimal error within 'line', and then
+// take the average over all 'subs'. Thus it returns 0 if each sub has
+// an exact match with one of the strings in 'line'.
+function subsMinError (subs, line) {
```
If this is an average, the name should be `subsAverageError` not `subsMinError`.

sanette replied:
let's do subsAvgMinError then

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+		words = text.split("  ");
+	    } else {
+		words = text.split(" ");
+	    };
```
Contextual separators sounds like a bad idea. I would propose to use `or` (since this is an OCaml keyword) or `|`.  Also is this feature used? Note that the search is currently inefficient for large number of alternatives (which seems fine?).

sanette replied:
what do you mean by inefficient (slow, or non-relevant results)?

I'm in favor of keeping the double space. It's not exactly a logical "or" because the ranking favors the match of both sides.
For instance if you want the signature 'int -> int' within the 'Int' module you can search "Int  int -> int"
Of course I have no idea whether this feature is used by others than me ;)

Octachron replied:
With the current code, for k words the input will be scanned k times in `hasSubstrings` which is unnecessary. For large number of words, it is probably better to build an or-regexp.
Assigning two different meanings to space depending on the presence of arrows is not a good idea. Similarly, the difference between ` ` and `  ` is far too small. I am not opposed to the feature but it should use a non-ambiguous token (`;` for instance?).

sanette replied:
> With the current code, for k words the input will be scanned k times in `hasSubstrings` which is unnecessary.

really? I was assuming that `subs.findIndex` returns as soon as a match is found

Octachron replied:
It does, but going from k to k/2 does not matter much for the asymptotic complexity.

sanette replied:
> Assigning two different meanings to space depending on the presence of arrows is not a good idea. Similarly, the difference between ` ` and ` ` is far too small. I am not opposed to the feature but it should use a non-ambiguous token (`;` for instance?).

I think it is important to keep a single space for the usual "google-like search" that everybody is now used to, so that if you type "print character" you get `Stdlib.print_char` as a result. However when searching for a type, necessarily the space acquires another meaning. But in most cases, my impression is that people will either search for a type or for keywords, but not both at the same time; then the problem disappears. If they really want to combine both, the fact that is becomes a bit tricky is not so scandalous, i think.

Octachron replied:
The google version would be to use `"..."` for searching types. Your solution seems far too brittle and ad hoc. If you want to optimize for single type search, I would propose a more structured alternative: words are grouped with `"..."`, and the toplevel search is first considered as a single group and only broken down to words if there is no search result.

sanette replied:
yes I was thinking of using double quotes too; let's do this

sanette replied:
> It does, but going from k to k/2 does not matter much for the asymptotic complexity.

right, but I think we are talking here about k=2,3, or 4 maximum, don't you think so?

Octachron replied:
I agree, thus the disclaimer that it is probably fine for the currently expected values of `k`.

sanette replied:
done (I mean "google style" search with quotes). I also did some benchmarks to optimize search speed a bit

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+    // sub = sub.toLowerCase();
+    // case sensitive is better for OCaml modules.
+    let ss = "^" + s + "$";
+    return (ss.indexOf(sub) !== -1);
```
This can be simplified to `ss.includes(sub)`, isn't it?

sanette replied:
@Octachron many thanks for the review. I am very busy this week, I hope to be able to work on this next week-end.

Octachron replied:
This is not an issue at all. I should have finished a full review by the end of the week. Thanks for all the work!

sanette replied:
> This can be simplified to `ss.includes(sub)`, isn't it?

yes thanks

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+	}
+	html += (' : ' + line[SIG_INDEX] + '</pre>');
+	html = '<pre>' + html + line[DESCR_INDEX]; }
+    return (li + html + "</li>\n");
```
I would suggest to use string interpolation in this function.

sanette replied:
good idea, thx

sanette committed:
use Option

sanette commented:
> Do you have some examples of online books using this kind of repetitions?

nope. Pure invention ;)

But I see what you mean. for long preambles, it can appear surprising. I'll try to follow your suggestion and replace them by the TOC; I think this also means removing the chapter title, then.
EDIT: maybe we can keep the chapter title, after all. Tell me what you prefer.

sanette committed:
https://github.com/ocaml/ocaml/pull/9755#discussion_r484764944

and
https://github.com/ocaml/ocaml/pull/9755#issuecomment-699508742
+
mobile navigation using some javascript

sanette commented:
@Octachron , @dra27 , I think I have implemented all your corrections and suggestions.

Moreover, following @Octachron 's question on mobile navigation, I have implemented some javascript menus to help mobile users, following standard practice (it should be similar to, for instance, the Julia Manual https://docs.julialang.org/en/v1/).
This lead to adding the "part title" on top of each page. While it's good in the mobile layout, I'm not completely sure it adds a value in the desktop layout. If not, I can hide it with the CSS, let me know what you think

sanette committed:
css for <ul> in main content

sanette committed:
correct title detection for index.html

sanette committed:
fix regression in <li> formatting

sanette commented:
I've uploaded (temporarily) the webman there:
https://sanette.github.io/ocaml-tutorial/webman/manual/index.html
in order to test it more easily

Octachron commented on manual/manual/html_processing/js/search.js:
```diff
@@ -0,0 +1,242 @@
+	    };
+	    words = words.filter(function (s) {
+		return (s !== "")});
+	    console.log (words);
```
The console logging should probably disappear.

sanette replied:
This was already removed, you're probably looking at an older version

sanette replied:
but there is another console.log `console.log("Search results = " + (count.toString()));`
which can be removed too if you want; although this one is only one integer, so not too invasive.

sanette committed:
search performance experiments

sanette committed:
cleanup search.js

sanette committed:
css

Octachron commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,465 @@
+      if count = 2 then begin
+        add_class "previous" (div $$ "a" |> R.first);
+        add_class "next" (div $$ "a" |> R.last);
+      end
```
The double fold seems too complicated. You could replace it by:

* fetch all previous, next, up, images
* filter the case 0 or 2 elements
* eliminate the version of the images at the top.
* Select 2 elements among the non-empty elements of the triplets
* Replace the first non-empty element with the new nav items.

sanette replied:
ok, I've simplified this along your lines. It's not shorter, but probably less convoluted indeed.

Octachron commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,465 @@
+  update_navigation xternal toc;
+  Option.iter delete (xternal $? "hr");
+  let xbody = xternal $ "body" in
+  let clone = parse (to_string soup) in
```
`parse(to_string soup)` should be avoided if possible. I would recommend to write a function to get the cut point (either the title or the h1 elements). Then writing to the filesystem can be done by mutating the next siblings of this cut point, then save to file.

sanette replied:
done

Octachron commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,465 @@
+  Re.Str.(global_replace (regexp ".+  ") "" s)
+
+let toc_get_title li =
+  let r = Re.Str.regexp ".+  " in
```
this the `remove_number` regexp, isn't it?

sanette replied:
good catch

Octachron commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,465 @@
+  List.iter (clone_structure soup toc_table) xfiles;
+
+  (* And finally save *)
+  save_to_file soup file
```
If you commute `save_to_file` and the generation of the external files, I have the impression that you don't need to deep copy the `soup` nodes.

sanette replied:
excellent point. This allows indeed the removal of all clones.

Octachron commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,380 @@
+   "<code class=\"type\">'a list -&gt; 'a list -&gt; 'a list</code>"; "'a list -> 'a list -> 'a list"]
+
+*)
+
```
I would propose to make the index generation a separated module.

sanette replied:
OK; this was actually the case in the original version, for the "fast" Index.make that we decided to remove.

sanette replied:
done

Octachron commented on manual/manual/html_processing/src/process_api.ml:
```diff
@@ -0,0 +1,380 @@
+  (* Add favicon *)
+  let head = soup $ "head" in
+  parse ({|<link rel="shortcut icon" type="image/x-icon" href="|} ^ favicon ^ {|">|})
+  |> append_child head;
```
The separation and top comments are a good idea. I would propose to go one step further and make all those parts of the processing pipeline a separated function.

sanette replied:
done

Octachron commented:
The mobile version works really well (nitpicking: in code examples, the result and the code should probably not have separated sliders). I had some last comments on the processing code and the travis test probably need to be updated. But overall, I think we should investigate using this new post-processing phase for the 4.12 manual.

sanette committed:
start review 9755#discussion_r509348495

=> simplify "update_navigation"
https://github.com/ocaml/ocaml/pull/9755#discussion_r509348495

=> download logo at proper
place (https://github.com/ocaml/ocaml.org/pull/1166#event-3927979984)

=> split pipeline into more functions
https://github.com/ocaml/ocaml.org/pull/1166#event-3927979984

sanette committed:
get rid of cloning

https://github.com/ocaml/ocaml/pull/9755#discussion_r509354188

sanette committed:
cleanup structure

move make_index to a separate Index module
https://github.com/ocaml/ocaml/pull/9755#discussion_r509355292

sanette committed:
move toc creation to a function

avsm commented:
Just a headsup that I've pushed @sanette's rearrangement of https://ocaml.org/docs/ live, so we should be set to add a beta version of the new docs there too.  There is a staging.ocaml.org now (synched from pushes to the `staging` branch of the website repo), so we can preview these docs there if you'd like once included in a beta release.

sanette commented:
> (nitpicking: in code examples, the result and the code should probably not have separated sliders)

done

sanette commented:
Maybe it would be a good idea to squash everything into one commit; I you don't oppose, I can do it

sanette committed:
add ocaml-re dependency for Travis CI

sanette committed:
add ocaml-re to .gitignore

sanette commented:
A question for @dra27 : you have added a .gitignore file to make the Travis CI happy, but wouldn't it be more logical to have the travis test call `make distclean`? I just proposed this PR: https://github.com/ocaml/ocaml/pull/10040 to this effect, but maybe that's the wrong place. What do you think?

dra27 commented:
`.gitignore` should always contain entries for things which should never committed - that's for while developing to ensure you don't commit files you shouldn't (with a stray `git add .` or, 😱, `git add -A`). The CI checks for the main distribution that `make distclean` definitely removes all non-committed files which is a hygiene check on the distclean target.

sanette commented:
> The CI checks for the main distribution that `make distclean` definitely removes all non-committed files which is a hygiene check on the distclean target.

yes but currently the CI does *not* call `make distclean` on the manual, but *still* checks left-overs, which sounds not logical to me.

dra27 commented:
> > The CI checks for the main distribution that `make distclean` definitely removes all non-committed files which is a hygiene check on the distclean target.
> 
> yes but currently the CI does _not_ call `make distclean` on the manual, but _still_ checks left-overs, which sounds not logical to me.

Indeed - it sounds as though the change doesn't need spinning off into a separate PR, you just want `make -C manual distclean` added to the Travis script here.

dra27 commented on manual/Makefile:
```diff
@@ -40,3 +43,7 @@ clean:
 	$(MAKE) -C tests  clean
+
+.PHONY: distclean
+distclean:
```
My bad I think - this should be:

```suggestion
distclean: clean
```

dra27 commented on manual/manual/Makefile:
```diff
@@ -165,8 +169,13 @@ clean:
 	cd infoman; rm -f ocaml.info ocaml.info-*  *.haux *.hind
+
+.PHONY: distclean
+distclean:
```
Likewise:
```suggestion
distclean: clean
```

dra27 reviewed (commented):
The `distclean` targets should depend on `clean` so that they trigger cleans right the way through. `make -C manual distclean` can be put below L104 here:
https://github.com/ocaml/ocaml/blob/4d6aba29f4835e8cd738e56b1ad2d3cf2cc782c9/tools/ci/actions/runner.sh#L99-L109

Octachron commented on manual/manual/html_processing/src/process_manual.ml:
```diff
@@ -0,0 +1,481 @@
+(* Create a new file by keeping only the head/headers parts of "soup", deleting
+   everything after the title, and inserting the content of external file (hence
+   preserving TOC and headers) (WARNING: this mutates soup) *)
+let clone_structure soup (title, header) toc xfile =
```
The name `clone_structure` does not seems to fit the notion of the function. Maybe `insert_xfile_content`?

sanette replied:
done

Octachron commented:
Once the clean issue is fixed, and this PR rebased; I propose to merge this PR and use the new manual format for the ocaml.org version of the 4.12 manual.

The conflicts are two changes: one in the manual's Makefile and one in the travis CI; they can be both discarded.

If I am not mistaken, all changes to the travis CI can be discarded too with the move to github actions. Am I wrong @dra27 ?

Since there was a lot of back-and-forth during the review process (and some transient binary artifacts), I am inclined to squash all commits, except if @sanette prefer to clean the history.

sanette committed:
insert_xfile_content

Octachron commented:
All that is missing now is a change entry, in the 4.12 manual section.

sanette committed:
'Changes' entry

gasche commented:
@sanette there is a conflict in the Changes file. You should probably rebase the PR on the current trunk.

sanette commented:
mmm, I tried to rebase on current trunk, I'm not sure I did the right thing though.

Well, it seems to work.

Octachron commented:
The first commit seems to be from an older version that used `manual/manual/docs_make`. You can drop it with `git rebase -i trunk`, and replacing `pick` by `drop` on the first line.

sanette commented:
> The first commit seems to be from an older version that used `manual/manual/docs_make`. You can drop it with `git rebase -i trunk`, and replacing `pick` by `drop` on the first line.

weird, how is this possible? which commit it this precisely? when i do `git rebase -i trunk` the first line is `pick 405fd3830 Missing .gitignore entries`

sanette commented:
do you want me to squash everything to one commit?

Octachron commented:
It is probably your local  `trunk` branch contains that contains this commit. (One way to solve this issue wcould be to switch to trunk with `git checkout trunk` and then reset the local `trunk` branch to the upstream version with `git reset your_name_for_the_upstream_repo/trunk`).

Octachron commented:
Squashing to one commit can be done directly on the github interface, so this is your choice. You could choose to go the extra-miles and clean-up the history to remove the back-and-forth (particularly for transient binaries), but contracting the history to one single commit seems fine to me here.

sanette commented:
> One way to solve this issue wcould be to switch to trunk with `git checkout trunk` and then reset the local `trunk` branch to the upstream version with `git reset your_name_for_the_upstream_repo/trunk`

I did this but it seems to be still there. Will try to fix this when I get the time

sanette commented:
ah. it should be better now.
It's ok with me if you squash everything when you merge. The only part that has disappeared during the review and that I found interesting to keep, just in case, is the fast index generator (1sec instead of 15sec). But this will stay on my github branch so that's fine.

Octachron commented:
Merged. Thanks a lot for your hard work and patience!
